connector/src/it/scala/org/apache/spark/sql/cassandra/execution/JoinThrottlingSpec.scala

```diff
@@ -0,0 +1,66 @@
+package org.apache.spark.sql.cassandra.execution
+
+import com.datastax.spark.connector.SparkCassandraITFlatSpecBase
+import com.datastax.spark.connector.cluster.DefaultCluster
+import com.datastax.spark.connector.cql.CassandraConnector
+import com.datastax.spark.connector.embedded.SparkTemplate
+import com.datastax.spark.connector.rdd.ReadConf
+import org.apache.spark.sql.cassandra.CassandraSourceRelation.DirectJoinSettingParam
+import org.apache.spark.sql.cassandra._
+import org.scalatest.concurrent.Eventually
+
+import scala.concurrent.Future
+
+class JoinThrottlingSpec extends SparkCassandraITFlatSpecBase with DefaultCluster with Eventually {
+
+  override lazy val conn = CassandraConnector(defaultConf)
+
+  private val rowsCount = 10000
+  private val rowsPerSecondPerCore = 400
+
+  override def beforeClass {
+    spark.conf.set(DirectJoinSettingParam.name, "auto")
+    spark.conf.set(ReadConf.ReadsPerSecParam.name, "400")
+
+    conn.withSessionDo { session =>
+      val executor = getExecutor(session)
+      createKeyspace(session)
+      awaitAll(
+        Future {
+          session.execute(s"CREATE TABLE IF NOT EXISTS $ks.kvtarget (k int PRIMARY KEY, v int, id int)")
+          session.execute(s"CREATE TABLE IF NOT EXISTS $ks.kv (k int PRIMARY KEY, v int)")
+          val ps = session.prepare(s"INSERT INTO $ks.kv (k,v) VALUES (?,?)")
+          awaitAll {
+            for (id <- 1 to rowsCount) yield {
+              executor.executeAsync(ps.bind(id: java.lang.Integer, id: java.lang.Integer))
+            }
+          }
+        }
+      )
+      executor.waitForCurrentlyExecutingTasks()
+    }
+  }
+
+  /* SPARKC-627 */
+  it should "throttle join by rows per second" in {
+    import spark.implicits._
+
+    val right = spark.range(1, rowsCount).map(_.intValue).withColumnRenamed("value", "id")
+    val left = spark.read.cassandraFormat("kv", ks.toLowerCase)
+      .option(ReadConf.ReadsPerSecParam.name, rowsPerSecondPerCore)
+      .load()
+    val join = left.join(right, left("k") === right("id"))
+
+    val startMillis = System.currentTimeMillis()
+    join.write.format("org.apache.spark.sql.cassandra")
+      .options(Map("keyspace" -> ks, "table" -> "kvtarget"))
+      .mode("append")
+      .save()
+    val durationSeconds = (System.currentTimeMillis() - startMillis).toInt / 1000
```

Comment:
Seems like a long test to check this, but logic seems sound to me

Reply:
Maybe put an assert in here that minimalDurationSeconds > 2 or 3 ? Just to make sure the test isn't doing nothing?

Reply:
Maybe just loop over a few thresholds and make sure each is slower than the previous? I dunno, ignore me if I'm just paranoid

Reply:
Thanks for the review :) It takes ~15 seconds on my laptop. We could lower this if this too long.

Reply:
My worry is that, what if unthrottled it takes 15 seconds (or greater) and so the test passes just because the CI environment is slow, not because of the throttle.

Reply:
I see, without throttling the test is an order of magnitude slower, around 2 seconds.

Reply:
But what about when I run the tests on my raspberry pi :) If you are sure it will be faster than I'm :+1: